src/coreclr/src/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.cs

```diff
@@ -0,0 +1,242 @@
+        /// <remarks>
+        /// Providing a <paramref name="wrapper"/> instance means <see cref="ComWrappers.GetOrCreateObjectForComInstance(IntPtr, CreateObjectFlags, object?)"/>
+        /// will not be called.
+        ///
+        /// If the <paramref name="wrapper"/> instance already has an associated external object a <see cref="System.NotSupportedException"/> will be thrown.
+        /// </remarks>
+        public object GetOrCreateObjectForComInstance(IntPtr externalComObject, CreateObjectFlags flags, object? wrapper = null)
+        {
+            if (externalComObject == IntPtr.Zero)
+                throw new ArgumentNullException(nameof(externalComObject));
+
+            ComWrappers impl = this;
+            object? wrapperLocal = wrapper;
+            object? retValue = null;
+            GetOrCreateObjectForComInstanceInternal(ObjectHandleOnStack.Create(ref impl), externalComObject, flags, ObjectHandleOnStack.Create(ref wrapperLocal), ObjectHandleOnStack.Create(ref retValue));
+
+            return retValue!;
+        }
+
+        [DllImport(RuntimeHelpers.QCall)]
+        private static extern void GetOrCreateObjectForComInstanceInternal(ObjectHandleOnStack comWrappersImpl, IntPtr externalComObject, CreateObjectFlags flags, ObjectHandleOnStack wrapper, ObjectHandleOnStack retValue);
+
+        /// <summary>
+        /// Create a managed object for the object pointed at by <paramref name="externalComObject"/> respecting the values of <paramref name="flags"/>.
+        /// </summary>
+        /// <param name="externalComObject">Object to import for usage into the .NET runtime.</param>
+        /// <param name="flags">Flags used to describe the external object.</param>
+        /// <returns>Returns a managed object associated with the supplied external COM object.</returns>
+        /// <remarks>
+        /// If the object cannot be created and <code>null</code> is returned, the call to <see cref="ComWrappers.GetOrCreateObjectForComInstance(IntPtr, CreateObjectFlags, object?)"/> will throw a <see cref="System.ArgumentNullException"/>.
+        /// </remarks>
+        protected abstract object? CreateObject(IntPtr externalComObject, CreateObjectFlags flags);
+
+        // Call to execute the abstract instance function
+        internal static object? CallCreateObject(ComWrappers? comWrappersImpl, IntPtr externalComObject, CreateObjectFlags flags)
+            => (comWrappersImpl ?? s_globalInstance!).CreateObject(externalComObject, flags);
+
+        /// <summary>
+        /// Called when a request is made for a collection of objects to be released.
+        /// </summary>
+        /// <param name="objects">Collection of objects to release.</param>
+        /// <remarks>
+        /// The default implementation of this function throws <see cref="System.NotImplementedException"/>.
+        /// </remarks>
+        protected virtual void ReleaseObjects(IEnumerable objects)
+        {
+            throw new NotImplementedException();
```

Comment:
Would it make more sense for the default impl to be no-op?

Reply:
I thought about that. My feeling was a request for releasing objects isn't something that should be ignorable. If there is nothing to do be done then implementers can state that explicitly. I want implementers to be put into a position to have to make decisions.

Reply:
The typical API design preference is to avoid forcing decision like this by providing a reasonable default implementation.

The class may not even be abstract. We can make the two abstract methods to have a reasonable default implementations. 

Something to ask during API review...